test/e2e/gardener/shoot/internal/inclusterclient/inclusterclient.go

```diff
@@ -0,0 +1,281 @@
+	By("Verify we are using the expected path")
+	Eventually(
+		execute(ctx, f.ShootClient, podName, "/kubectl", "cluster-info"),
+	).Should(Say(
+		"Kubernetes control plane is running at %s", expectedAddress,
+	))
+
+	By("Verify a typical API request works")
+	Eventually(
+		execute(ctx, f.ShootClient, podName, "/kubectl", "get", "pod", podName),
+	).Should(Say(
+		podName,
+	))
+}
+
+func getInternalAPIServerAddress(shoot *gardencorev1beta1.Shoot) string {
+	GinkgoHelper()
+
+	var address string
+	for _, a := range shoot.Status.AdvertisedAddresses {
+		if a.Name == v1beta1constants.AdvertisedAddressInternal {
+			address = a.URL
+			break
+		}
+	}
+	Expect(address).NotTo(BeEmpty())
+
+	return address + ":443"
+}
+
+func getInClusterAPIServerAddress(ctx context.Context, c client.Client) string {
+	GinkgoHelper()
+
+	service := corev1.Service{}
+	Expect(c.Get(ctx, client.ObjectKey{Name: "kubernetes", Namespace: metav1.NamespaceDefault}, &service)).To(Succeed())
+
+	clusterIP := service.Spec.ClusterIP
+	Expect(clusterIP).NotTo(BeEmpty())
+
+	var port int32
+	for _, p := range service.Spec.Ports {
+		if p.Name == "https" {
+			port = p.Port
+			break
+		}
+	}
+	Expect(port).NotTo(BeZero())
+
+	return fmt.Sprintf("https://%s:%d", clusterIP, port)
+}
+
+func prepareObjects(ctx context.Context, c client.Client, objects ...client.Object) {
+	By("Create test objects for verifying in-cluster access to API server")
+	for _, obj := range objects {
+		Expect(client.IgnoreAlreadyExists(c.Create(ctx, obj))).To(Succeed(), "should create %T %q", obj, client.ObjectKeyFromObject(obj))
+	}
+
+	DeferCleanup(func() {
+		By("Cleaning up test objects for verifying in-cluster access to API server")
+		cleanupCtx, cancel := context.WithTimeout(context.Background(), time.Minute)
+		defer cancel()
+
+		for _, obj := range objects {
+			Expect(client.IgnoreNotFound(c.Delete(cleanupCtx, obj))).To(Succeed(), "should delete %T %q", obj, client.ObjectKeyFromObject(obj))
+		}
+	})
+
+	By("Wait for test pods to be ready")
+	for _, obj := range objects {
+		pod, ok := obj.(*corev1.Pod)
+		if !ok {
+			continue
+		}
+
+		Eventually(func(g Gomega) {
+			g.Expect(c.Get(ctx, client.ObjectKeyFromObject(pod), pod)).To(Succeed())
+			g.Expect(pod.Status.Phase).To(Equal(corev1.PodRunning))
+		}).WithContext(ctx).WithTimeout(time.Minute).Should(Succeed(), "%T %q should get ready", obj, client.ObjectKeyFromObject(obj))
+	}
+}
+
+func execute(ctx context.Context, clientSet kubernetes.Interface, podName string, command ...string) *Buffer {
+	GinkgoHelper()
+	var stdOutBuffer, stdErrBuffer = NewBuffer(), NewBuffer()
+
+	Expect(clientSet.PodExecutor().ExecuteWithStreams(
+		ctx,
+		namespace,
+		podName,
+		containerName,
+		nil,
+		io.MultiWriter(stdOutBuffer, gexec.NewPrefixedWriter("[out] ", GinkgoWriter)),
+		io.MultiWriter(stdErrBuffer, gexec.NewPrefixedWriter("[err] ", GinkgoWriter)),
+		command...,
+	)).To(Succeed())
+
+	return stdOutBuffer
+}
```

Comment:
Will it cause headaches later on that we basically ignore `stderr` here? In case anything is ever returned in `stderr` it will not change the test result or be visible anywhere. Not sure if it will be helpful if something went wrong to just have the information that the execute statement failed when it actually provided some feedback.

Reply:
I explicitly tested what happens if the `kubectl` command fails.
In this case, `ExecuteWithStreams` returns an error that the command returned a non-zero exit code, which causes the test to fail.
As both streams are forwarded to the ginkgo output (prefixed with `[out]` and `[err]`), it's easy to understand what caused the command to fail.